### ISSUE TITLE
SignalR.Client 2.2.2 Start fails with NullReferenceException when setting user-agent header

### ISSUE THREAD

**Issue by AlexanderMelchers** [SignalR/SignalR#3936]
### Expected  behavior
When I call the `HubConnection.Start()`-method, the SignalR-client starts and is able to receive push notifications from the server.

### Actual behavior
The `HubConnection.Start()`-method raises a `NullReferenceException`.

This only happens when setting a user-agent on the `HubConnection`, however, using a call to `HubConnection.Headers.Add(string, string)`, and only when using library version 2.2.2. It **does not occur** when using version 2.2.1 (verified this as the problem did not exist before moving to 2.2.2, and disappeared following a downgrade back to 2.2.1).

### Steps to reproduce
1. Create a new hub connection:
 `HubConnection connection = new HubConnection(url, false);`
2. Set it's cookie container (this may or may not be relevant):
 `connection.CookieContainer = this.CookieContainer;`
3. Set the connection's user-agent:
 `connection.Headers.Add("User-Agent", "My valid user agent");`
4. Register subscriptions:
 `this.Subscriptions = new[] { this.Proxy.On<MessageType>(nameof(this.MessageEvent), message => this.MessageEvent?.Invoke(this, message)) };`
5. Start the SignalR client connection:
`await this.Connection.Start().ContinueWith(task => { this.IsActive = true; this.OnStarted(); }, TaskContinuationOptions.OnlyOnRanToCompletion).ConfigureAwait(false);`

The `NullReferenceException` is raised in step 5.

### Stack trace
The stack trace is as follows (as reported by HockeyApp):

> System.Net.Http.Headers.HeaderInfo.HeaderTypeInfo<T, U>.AddToCollection(object collection, object value)<c7b8b5731cb3482ba2d655eb25ee0f74>:0
System.Net.Http.Headers.HttpHeaders.AddInternal(string name, IEnumerable<T> values, HeaderInfo headerInfo, bool ignoreInvalid)<c7b8b5731cb3482ba2d655eb25ee0f74>:0
System.Net.Http.Headers.HttpHeaders.Add(string name, IEnumerable<T> values)<c7b8b5731cb3482ba2d655eb25ee0f74>:0
System.Net.Http.Headers.HttpHeaders.Add(string name, string value)<c7b8b5731cb3482ba2d655eb25ee0f74>:0
Microsoft.AspNet.SignalR.Client.Http.HttpRequestMessageWrapper.SetRequestHeaders(IDictionary<TKey, TValue> headers)<14fb36c86d16485e985f83bccff407db>:0
Microsoft.AspNet.SignalR.Client.Connection.Microsoft.AspNet.SignalR.Client.IConnection.PrepareRequest(IRequest request)<14fb36c86d16485e985f83bccff407db>:0
Microsoft.AspNet.SignalR.Client.Http.DefaultHttpClient.Get(string url, Action<T> prepareRequest, bool isLongRunning)<14fb36c86d16485e985f83bccff407db>:0
Microsoft.AspNet.SignalR.Client.Transports.TransportHelper.GetNegotiationResponse(IHttpClient httpClient, IConnection connection, string connectionData)<14fb36c86d16485e985f83bccff407db>:0
Microsoft.AspNet.SignalR.Client.Transports.AutoTransport.GetNegotiateResponse(IConnection connection, string connectionData)<14fb36c86d16485e985f83bccff407db>:0
Microsoft.AspNet.SignalR.Client.Transports.AutoTransport.Negotiate(IConnection connection, string connectionData)<14fb36c86d16485e985f83bccff407db>:0
Microsoft.AspNet.SignalR.Client.Connection.Negotiate(IClientTransport transport)<14fb36c86d16485e985f83bccff407db>:0
Microsoft.AspNet.SignalR.Client.Connection.Start(IClientTransport transport)<14fb36c86d16485e985f83bccff407db>:0
Microsoft.AspNet.SignalR.Client.Connection.Start(IHttpClient httpClient)<14fb36c86d16485e985f83bccff407db>:0
Microsoft.AspNet.SignalR.Client.Connection.Start()<14fb36c86d16485e985f83bccff407db>:0
My.SignalR.SignalRClient.<StartAsync>d__23.MoveNext()<63e232f22232434f968a04e4632aca4c>:0

### Environment
The project set-up is a Xamarin.Forms project with both Android and iOS projects in the same solution, supported by PCLs for cross-platform code. The SignalR client is mainly used from within the PCLs, but added as a reference to the final output projects as well. Two different clients have been implemented, for two different hubs, and both have their own classes within the solution file. These classes share a common base class which has been abstracted into a NuGet-package.

The exception was experienced from within the Android project, but I expect it to also occur on iOS. Development takes place in Visual Studio Professional 2015 Update 3.

### Additional information
* The problem first occurred after upgrading SignalR.Client from version 2.2.1 to 2.2.2.
* The problem disappears after downgrading back to SignalR.Client version 2.2.1.
* The problem does not occur when the user-agent header is not added (step 3 above).
* Even though the stack trace and former bullet-point seem to indicate that the problem lies with the `System.Net.Http.Headers.HttpHeaders`-class, a test-case executed in a different project (though also Android) appears to indicate that there's no problem adding our user-agent string to a headers class instance. The steps taken to verify this are as follows:
 1. Look for the parameterless constructor of the `HttpRequestHeaders`-class using reflection (as you can;t otherwise instantiate an instance of this class):
 `ConstructorInfo ctor =typeof(HttpRequestHeaders).GetConstructors(BindingFlags.Instance | BindingFlags.NonPublic).FirstOrDefault(c => !c.GetParameters().Any());`
 2. Instantiate a new class instance:
 `HttpRequestHeaders headers = (HttpRequestHeaders) ctor.Invoke(new object[] {});`
 3. Set HTTP request headers as SignalR.Client does itself:
 `headers.TryAddWithoutValidation("User-Agent", "SignalR.Client.NetStandard/2.2.2 (Android)");`
4. Now add our custom user-agent headers:
 `headers.Add("User-Agent", "COOL/0.1 (Android 6.0; Marshmallow, API 23; Samsung; rv:34)");`

**Comment by maltium** [comment]
I confirm this issue when running on Mono

**Comment by AlexanderMelchers** [comment]
Just for those interested, this is still an issue in version 2.2.3.

**Comment by AlexanderMelchers** [comment]
Based on the stack trace I took a look at the code, starting at that place in the code that seemed most reasonable to me as the starting point from where header operations are begun, namely the [`Microsoft.AspNet.SignalR.Client.Http.HttpRequestMessageWrapper` -class](https://github.com/SignalR/SignalR/blob/dev/src/Microsoft.AspNet.SignalR.Client/Http/HttpWebRequestWrapper.cs). Sure enough at the top of this class there's a dictionary called `_restrictedHeadersSet` that gave me an idea of something I could still try to figure this thing out and resolve it... For though in this dictionary user-agent headers are allowed, they are not written in the usual RFC "User-Agent"-format, but as a custom "UserAgent" (a `ToString` of an enumeration value). As a consequence, setting a "User-Agent"-header is no longer allowed as of version 2.2.2, but setting a "UserAgent"-header is.

For me, then, the solution to the above issue is to change all places where I set a custom user-agent so that the user-agent is specified as "UserAgent". The SignalR client seems to work perfectly after that...
